pkg/graveler/ref/merge_base_finder.go

```diff
@@ -0,0 +1,60 @@
+package ref
+
+import (
+	"container/heap"
+	"context"
+
+	"github.com/treeverse/lakefs/pkg/graveler"
+)
+
+// taken from history:
+// https://github.com/treeverse/lakeFS/blob/606bf07969c14a569a60efe9c92831f424fa7f36/index/dag/commit_iterator.go
+type CommitGetter interface {
+	GetCommit(ctx context.Context, repositoryID graveler.RepositoryID, commitID graveler.CommitID) (*graveler.Commit, error)
+}
+
+type reachedFlags struct {
+	fromLeft  bool
+	fromRight bool
+}
+
+func (d reachedFlags) or(fromLeft bool, fromRight bool) reachedFlags {
+	d.fromLeft = d.fromLeft || fromLeft
+	d.fromRight = d.fromRight || fromRight
+	return d
+}
```

Comment:
If you use bitmask flags it can simplify some of the code
```suggestion
type ReachedFlags uint8

const (
	ReachedLeft ReachedFlags = 1 << iota
	ReachedRight
)

func (r ReachedFlags) ReachedBoth() bool {
	return r&ReachedRight != 0 && r&ReachedLeft != 0
}
```

Reply:
Done